src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferencesSearchEngine.SymbolSet.cs

```diff
@@ -0,0 +1,276 @@
+        /// that clients will also need.  This includes:
+        /// <list type="number">
+        /// <item>Cascading to all linked symbols for the requested symbol.  This ensures a unified set of results for a
+        /// particular symbol, regardless of what project context it was originally found in.</item>
+        /// <item>Symbol specific cascading.  For example, when searching for a named type, references to that named
+        /// type will be found through its constructors.</item>
+        /// <item>Cascading up and down the inheritance hierarchy for members (e.g. methods, properties, events). This
+        /// is controllable through the <see cref="FindReferencesSearchOptions.UnidirectionalHierarchyCascade"/>
+        /// option.</item>
+        /// </list>
+        /// </summary>
+        private abstract class SymbolSet
+        {
+            protected readonly FindReferencesSearchEngine Engine;
+
+            protected SymbolSet(FindReferencesSearchEngine engine)
+            {
+                Engine = engine;
+            }
+
+            protected Solution Solution => Engine._solution;
+
+            /// <summary>
+            /// Get a copy of all the symbols in the set.  Cannot be called concurrently with <see
+            /// cref="InheritanceCascadeAsync"/>
+            /// </summary>
+            public abstract ImmutableArray<ISymbol> GetAllSymbols();
+
+            /// <summary>
+            /// Update the set of symbols in this set with any appropriate symbols in the inheritance hierarchy brought
+            /// in within <paramref name="project"/>.  For example, given a project 'A' with interface <c>interface IGoo
+            /// { void Goo(); }</c>, and a project 'B' with class <c>class Goo : IGoo { public void Goo() { } }</c>,
+            /// then initially the symbol set will only contain IGoo.Goo.  However, when project 'B' is processed, this
+            /// will add Goo.Goo is added to the set as well so that references to it can be found.
+            /// </summary>
+            /// <remarks>
+            /// This method is non threadsafe as it mutates the symbol set instance.  As such, it should only be
+            /// called serially.  <see cref="GetAllSymbols"/> should not be called concurrently with this.
+            /// </remarks>
+            public abstract Task InheritanceCascadeAsync(Project project, CancellationToken cancellationToken);
+
+            private static bool InvolvesInheritance(ISymbol symbol)
+                => symbol.Kind is SymbolKind.Method or SymbolKind.Property or SymbolKind.Event;
+
+            protected static void PushAll(Stack<ISymbol> stack, ImmutableArray<ISymbol> symbols)
+            {
+                foreach (var symbol in symbols)
+                    stack.Push(symbol);
+            }
+
+            protected static void PushAll(Stack<ISymbol> stack, HashSet<ISymbol> symbols)
```

Comment:
Couldn't these just be extension methods?

Reply:
yup. moved!